lib/bandit/io.ex

```diff
@@ -0,0 +1,55 @@
+defmodule Bandit.IO do
+  @moduledoc false
+
+  @doc """
+  Returns an iolist consisting of the first `length` bytes of the given list, along with the
+  remainder of the iolist beyond it as a tuple. Returns `{:error, :length_exceeded}` if there are
+  not `length` bytes remaining in the iolist.
+  """
+  @spec split_iodata(iodata(), pos_integer(), iodata(), iodata()) :: {iodata(), iodata()} | {:error, :length_exceeded}
+  def split_iodata(iodata, length, head \\ [], acc \\ [])
+  def split_iodata(iodata, 0, [] = head, []), do: {head, iodata}
+  def split_iodata([], _length, [], []), do: {:error, :length_exceeded}
+  def split_iodata("", _length, [], []), do: {:error, :length_exceeded}
+  def split_iodata([byte | _] = iodata, 0, head, []) when is_integer(byte), do: {head, iodata}
+  def split_iodata(_iodata, 0, head, [] = acc), do: {head, acc}
+  def split_iodata(iodata, 0, head, acc), do: {head, [iodata | acc]}
+  def split_iodata([], _length, head, acc), do: {head, acc}
+  def split_iodata([[] | iodata], length, head, acc), do: split_iodata(iodata, length, head, acc)
+  def split_iodata(["" | iodata], length, head, acc), do: split_iodata(iodata, length, head, acc)
+  def split_iodata([byte | iodata], length, head, acc) when is_integer(byte), do: split_iodata(iodata, length - 1, head ++ [byte], acc)
+  def split_iodata([h | iodata], length, head, [] = acc) when h != [] do
```

Comment:
Is pattern-matching preferable over guards when you can do it? (Or does it not matter?)

```suggestion
  def split_iodata([[_ | _] = h | iodata], length, head, acc) do
```

Reply:
Good question, I'm not really sure but it wouldn't surprise me if it already got optimized away.